### PR TITLE
Support `Refinement` in `Preficate.tuple` and `Predicate.struct`

### PR DESCRIPTION
**Pull request by KhraksMamtsov**
support combination predicates and refinements

## Type

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Related

- Related Issue #
- Closes #

### CODE REVIEW

File: packages/effect/src/Predicate.ts [hunk 1]
```diff
@@ -686,31 +725,51 @@ export const productMany = <A>(
...
  *
  * @since 2.0.0
  */
-export const tuple = <T extends ReadonlyArray<Predicate<any>>>(
-  ...elements: T
-): Predicate<Readonly<{ [I in keyof T]: [T[I]] extends [Predicate<infer A>] ? A : never }>> => all(elements) as any
+export const tuple: {
+  <T extends ReadonlyArray<Predicate.Any>>(
+    ...elements: T
+  ): true extends { [K in keyof T]: T[K] extends Refinement.Any ? true : false }[number] ? Refinement<
```
Comment (tim-smart) [inline packages/effect/src/Predicate.ts:738]: Would this be better? Then you can remove the intermediate record:

```ts
[Extract<A[keyof A], Predicate.Refinement<any, any>>] extends [never]
```

File: packages/effect/src/Predicate.ts [hunk 2]
```diff
@@ -686,31 +725,51 @@ export const productMany = <A>(
...
+      Readonly<{ [I in keyof T]: T[I] extends Refinement.Any ? Refinement.In<T[I]> : Predicate.In<T[I]> }>,
```
Comment (tim-smart) [inline packages/effect/src/Predicate.ts:739]:
```suggestion
      { readonly [I in keyof T]: T[I] extends Refinement.Any ? Refinement.In<T[I]> : Predicate.In<T[I]> },
```

You can do it in one pass.

File: packages/effect/src/Predicate.ts (hunk 1 above)
Comment (KhraksMamtsov) [reply packages/effect/src/Predicate.ts]: Yes, It works and more elegant.

File: packages/effect/src/Predicate.ts (hunk 2 above)
Comment (KhraksMamtsov) [reply packages/effect/src/Predicate.ts]: done

**Pull request merged**